test/test_transformers.py

```diff
@@ -722,6 +723,62 @@ def set_weights_deterministic(model):
         self.assertEqual(result.shape, ref_output.shape)
         torch.testing.assert_close(result, ref_output, atol=1e-3, rtol=1e-2)
 
+    def sdp_ref(self,
+                q,
+                k,
+                v,
+                attn_mask=None,
+                dropout_p=0.0):
+        E = q.size(-1)
+        q = q / math.sqrt(E)
+        # (B, Nt, E) x (B, E, Ns) -> (B, Nt, Ns)
+        if attn_mask is not None:
+            attn = torch.baddbmm(attn_mask, q, k.transpose(-2, -1))
+        else:
+            attn = torch.bmm(q, k.transpose(-2, -1))
+
+        attn = torch.nn.functional.softmax(attn, dim=-1)
+        if dropout_p > 0.0:
```

Comment:
btw I think inside dropout there's also a shortcut for p == 0.0, but not 100% sure if it incurs a copy or not